cmd/query/app/query_parser.go

```diff
@@ -143,36 +193,184 @@ func (p *queryParser) parse(r *http.Request) (*traceQueryParameters, error) {
 	}
-	micros, err := strconv.ParseInt(value, 10, 64)
+	t, err := strconv.ParseInt(formValue, 10, 64)
 	if err != nil {
-		return time.Time{}, err
+		return time.Time{}, newParseError(err, paramName)
 	}
-	return time.Unix(0, 0).Add(time.Duration(micros) * time.Microsecond), nil
+	return time.Unix(0, 0).Add(time.Duration(t) * units), nil
 }
 
-func (p *queryParser) parseDuration(durationParam string, r *http.Request) (time.Duration, error) {
-	durationInput := r.FormValue(durationParam)
-	if len(durationInput) > 0 {
-		duration, err := time.ParseDuration(durationInput)
-		if err != nil {
-			return 0, fmt.Errorf("cannot not parse %s: %w", durationParam, err)
+// parseDuration parses the duration parameter of an HTTP request using the provided durationParser.
+// If the duration parameter is empty, the given defaultDuration will be returned.
+func parseDuration(r *http.Request, paramName string, parse durationParser, defaultDuration time.Duration) (time.Duration, error) {
+	formValue := r.FormValue(paramName)
+	if formValue == "" {
+		return defaultDuration, nil
+	}
+	d, err := parse(formValue)
+	if err != nil {
+		return 0, newParseError(err, paramName)
+	}
+	return d, nil
+}
+
+func parseBool(r *http.Request, paramName string) (b bool, err error) {
+	formVal := r.FormValue(paramName)
+	if formVal == "" {
+		return false, nil
+	}
+	b, err = strconv.ParseBool(formVal)
+	if err != nil {
+		return b, newParseError(err, paramName)
+	}
+	return b, nil
+}
+
+// parseSpanKindParam parses the input span kinds to filter for in the metrics query.
+//
+// Valid input span kinds include:
+// - "unspecified": when no span kind specified in span.
+// - "internal": internal operation within an application, instead of application boundaries.
+// - "server": server-side handling span.
+// - "client": outbound service call span.
+// - "producer": producer sending a message to broker.
+// - "consumer": consumer consuming a message from a broker.
+//
+// The output span kinds are the string representations from the OpenTelemetry model/proto/metrics/otelspankind.proto.
+// That is, the following map to the above valid inputs:
+// - "SPAN_KIND_UNSPECIFIED"
+// - "SPAN_KIND_INTERNAL"
+// - "SPAN_KIND_SERVER"
+// - "SPAN_KIND_CLIENT"
+// - "SPAN_KIND_PRODUCER"
+// - "SPAN_KIND_CONSUMER"
+func parseSpanKinds(r *http.Request, paramName string, defaultSpanKinds []string) ([]string, error) {
+	query := r.URL.Query()
+	jaegerSpanKinds, ok := query[paramName]
+	if !ok {
+		return defaultSpanKinds, nil
+	}
+	otelSpanKinds, err := mapSpanKinds(jaegerSpanKinds)
+	if err != nil {
+		return defaultSpanKinds, newParseError(err, paramName)
+	}
+	return otelSpanKinds, nil
+}
+
+func mapSpanKinds(spanKinds []string) ([]string, error) {
+	otelSpanKinds := make([]string, len(spanKinds))
+	var ok bool
+	for i, spanKind := range spanKinds {
+		if otelSpanKinds[i], ok = jaegerToOtelSpanKind[spanKind]; !ok {
+			return otelSpanKinds, fmt.Errorf("unsupported span kind: '%s'", spanKind)
```

Comment:
nit (more narrow var scope):
```suggestion
	for i, spanKind := range spanKinds {
		if v, ok := jaegerToOtelSpanKind[spanKind]; ok {
		    otelSpanKinds[i] = v
		} else {
			return otelSpanKinds, fmt.Errorf("unsupported span kind: '%s'", spanKind)
```